File: BotsSDK/korebot/src/main/java/kore/botssdk/adapter/AdvanceTableListAdapter.java
package kore.botssdk.adapter;

import static android.view.View.GONE;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Typeface;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;

import kore.botssdk.R;
import kore.botssdk.models.AdvanceListTableModel;
import kore.botssdk.utils.StringUtils;
import kore.botssdk.view.viewUtils.RoundedCornersTransform;

public class AdvanceTableListAdapter extends RecyclerView.Adapter<AdvanceTableListAdapter.ButtonViewHolder> {
    private final LayoutInflater inflater;
    private final ArrayList<AdvanceListTableModel.AdvanceTableRowDataModel> buttons;

    public AdvanceTableListAdapter(@NonNull Context context, @NonNull ArrayList<AdvanceListTableModel.AdvanceTableRowDataModel> buttons) {
        this.buttons = buttons;
        this.inflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public ButtonViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ButtonViewHolder(inflater.inflate(R.layout.advance_table_list_cell, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ButtonViewHolder holder, int i) {

        AdvanceListTableModel.AdvanceTableRowDataModel btn = buttons.get(i);
        holder.botListItemTitle.setText(btn.getTitle());

        if (!StringUtils.isNullOrEmpty(btn.getIcon()))
        {
            holder.botListItemImage.setVisibility(View.VISIBLE);
            RelativeLayout.LayoutParams layoutParams = new RelativeLayout.LayoutParams(100, 100);
            layoutParams.setMargins(0, 0, 10, 0);

            if(!StringUtils.isNullOrEmpty(btn.getIconSize()))
            {
                if(btn.getIconSize().equalsIgnoreCase("large"))
                {
                    layoutParams.height = 180;
                    layoutParams.width = 180;
                }
                else if(btn.getIconSize().equalsIgnoreCase("small"))
                {
                    layoutParams.height = 60;
                    layoutParams.width = 60;
                }
            }
            holder.botListItemImage.setLayoutParams(layoutParams);

            try {
                String imageData;
                imageData = btn.getIcon();
                if (imageData.contains(",")) {
                    imageData = imageData.substring(imageData.indexOf(",") + 1);
                    byte[] decodedString = Base64.decode(imageData.getBytes(), Base64.DEFAULT);
                    Bitmap decodedByte = BitmapFactory.decodeByteArray(decodedString, 0, decodedString.length);
                    holder.botListItemImage.setImageBitmap(decodedByte);
                } else {
                    Picasso.get().load(btn.getIcon()).transform(new RoundedCornersTransform()).into(holder.botListItemImage);
                }
            } catch (Exception e) {
                holder.botListItemImage.setVisibility(GONE);
            }
        }

        holder.botListItemTitle.setTypeface(null, Typeface.BOLD);
        if (!StringUtils.isNullOrEmpty(btn.getDescription())) {
            holder.botListItemSubtitle.setVisibility(View.VISIBLE);
            holder.botListItemSubtitle.setText(btn.getDescription());
        }
    }

    @Override
    public int getItemCount() {
        return buttons != null ? buttons.size() : 0;
    }

    boolean isFullView;

    public void setIsFromFullView(boolean isFullView) {
        this.isFullView = isFullView;
    }

    public static class ButtonViewHolder extends RecyclerView.ViewHolder {
        final TextView botListItemTitle;
        final TextView botListItemSubtitle;
        final ImageView botListItemImage;

        public ButtonViewHolder(@NonNull View itemView) {
            super(itemView);
            botListItemTitle = itemView.findViewById(R.id.bot_list_item_title);
            botListItemSubtitle = itemView.findViewById(R.id.bot_list_item_subtitle);
            botListItemImage = itemView.findViewById(R.id.bot_list_item_image);
        }
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/adapters/AdvanceTableListAdapter.kt
package com.kore.ui.adapters

import android.content.Context
import android.graphics.Typeface
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.kore.common.utils.BitmapUtils
import com.kore.model.constants.BotResponseConstants
import com.kore.ui.R

class AdvanceTableListAdapter(val context: Context, private val tables: List<Map<String, *>>) :
    RecyclerView.Adapter<TableViewDataHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TableViewDataHolder {
        return TableViewDataHolder(LayoutInflater.from(context).inflate(R.layout.advance_table_list_cell, parent, false))
    }

    override fun getItemCount(): Int {
        return tables.size
    }

    override fun onBindViewHolder(holder: TableViewDataHolder, position: Int) {
        val table = tables[position]

        if (table[BotResponseConstants.KEY_TITLE] != null) holder.botListItemTitle.text =
            table[BotResponseConstants.KEY_TITLE] as String

        if (table[BotResponseConstants.ICON] != null) {
            holder.botListItemImage.visibility = View.VISIBLE
            val layoutParams = RelativeLayout.LayoutParams(100, 100)
            layoutParams.setMargins(0, 0, 10, 0)
            if (table[BotResponseConstants.ICON_SIZE] != null) {
                when (table[BotResponseConstants.ICON_SIZE]) {
                    BotResponseConstants.LARGE -> {
                        layoutParams.height = 180
                        layoutParams.width = 180
                    }

                    BotResponseConstants.SMALL -> {
                        layoutParams.height = 60
                        layoutParams.width = 60
                    }
                }
            }
            holder.botListItemImage.layoutParams = layoutParams

            try {
                val imageData = table[BotResponseConstants.ICON] as String
                BitmapUtils.loadImage(context, imageData, holder.botListItemImage, R.drawable.ic_image_photo)
            } catch (e: Exception) {
                holder.botListItemImage.visibility = View.GONE
            }
        }

        holder.botListItemTitle.setTypeface(null, Typeface.BOLD)
        if (table[BotResponseConstants.DESCRIPTION] != null) {
            holder.botListItemSubtitle.visibility = View.VISIBLE
            holder.botListItemSubtitle.text = table[BotResponseConstants.DESCRIPTION] as String
        }
    }
}

class TableViewDataHolder(view: View) : RecyclerView.ViewHolder(view) {
    val botListItemTitle: TextView = itemView.findViewById(R.id.bot_list_item_title)
    val botListItemSubtitle: TextView = itemView.findViewById(R.id.bot_list_item_subtitle)
    val botListItemImage: ImageView = itemView.findViewById(R.id.bot_list_item_image)
}